File: ts/src/clock.ts
export interface Clock {
  now(): Date;
}

export class RealClock implements Clock {
  now(): Date {
    return new Date();
  }
}

export class ManualClock implements Clock {
  private _now: Date;

  constructor(now: Date = new Date(0)) {
    this._now = new Date(now.valueOf());
  }

  now(): Date {
    return new Date(this._now.valueOf());
  }

  set(now: Date): void {
    this._now = new Date(now.valueOf());
  }

  advance(ms: number): Date {
    this._now = new Date(this._now.valueOf() + ms);
    return this.now();
  }
}

File: py/src/apptheory/clock.py
from __future__ import annotations

import datetime as dt
from dataclasses import dataclass
from typing import Protocol


class Clock(Protocol):
    def now(self) -> dt.datetime: ...


@dataclass(slots=True)
class RealClock:
    def now(self) -> dt.datetime:
        return dt.datetime.now(tz=dt.UTC)


@dataclass(slots=True)
class ManualClock:
    _now: dt.datetime

    def __init__(self, now: dt.datetime | None = None) -> None:
        self._now = now or dt.datetime.fromtimestamp(0, tz=dt.UTC)

    def now(self) -> dt.datetime:
        return self._now

    def set(self, now: dt.datetime) -> None:
        self._now = now

    def advance(self, delta: dt.timedelta) -> dt.datetime:
        self._now = self._now + delta
        return self._now
